### ISSUE TITLE
Permission Denied (Docker Instance)

### ISSUE THREAD

**Issue by ajhorvath20** [mmgoodnow/cross-seed#62]
Hi, I just setup this but I am getting errors and I am sure its something to do with the way I am running the test. I have cross-see running on machine a and qbittorrent is running on machine b (definitely machine a can't see machine a shares). 

This is my compose 

```
cross-seed:
    image: mmgoodnow/cross-seed
    hostname: ${DOCKERHOSTNAME}
    ports:
      - 2468:2468
    container_name: cross-seed
    user: 1000:1000
    environment:
      - TZ=${TZ}
    restart: unless-stopped
    command: daemon
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - ${DOCKERCONFDIR}/cross-seed:/config
      - ${DOCKERSTORAGEDIR}/data/.torrents:/output
```

I am getting these errors

```
Configuration file already exists.
Validating your configuration...
Your configuration is valid!
(node:1) UnhandledPromiseRejectionWarning: Error: EACCES: permission denied, mkdir '/storage/data/.torrents'
    at Object.mkdirSync (fs.js:921:3)
    at Object.serve (/usr/src/app/dist/src/server.js:91:18)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async Command.<anonymous> (/usr/src/app/dist/src/cmd.js:98:13)
    at async Promise.all (index 0)
    at async run (/usr/src/app/dist/src/cmd.js:132:5)
(node:1) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:1) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

Thats my config.js file which I created manually

```module.exports = {
	configVersion: 1,

	jackettServerUrl: "http://192.168.1.167:9117",
	jackettApiKey: "apikey",
	delay: 10,
	trackers: ["oink", "tehconnection"],
	torrentDir: "/storage/data/.torrents",
	outputDir: "/storage/data/.torrents",
	includeEpisodes: false,
	searchAll: false,
	action: "save",
	rtorrentRpcUrl: undefined,
	qbittorrentUrl: "http://username:password@192.168.1.167:8080",
};
```
But I have that directory local where I am running cross-seed `/storage/data/.torrents`

**Comment by ajhorvath20** [comment]
Hi, managed to fix this by adding
```
    environment:
      - PGID=${PGID}
      - PUID=${PUID}
```
Your configuration is valid!
Server is running on port 2468, ^C to stop.

I have no idea why not `user: 1000:1000`

I can also see the `cross-seed` tag added in `qbittorrent` but I cannot see anything added in the `qbittorrent`. Am I missing anything?

Thanks

**Comment by mmgoodnow** [comment]
This looks like it could be a Docker path mapping issue. See config.template.docker:js:

https://github.com/mmgoodnow/cross-seed/blob/master/src/config.template.docker.js#L19-L30

This is the second report I've gotten of this type, so there very well could be something going on on my end. I'll gladly fix it if so.

**Comment by ajhorvath20** [comment]
Apologise when you say set the volume mapping you mean like that?
```
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - ${DOCKERCONFDIR}/cross-seed:/config
      - ${DOCKERCONFDIR}/cross-seed:/cross-seeds
      - ${DOCKERCONFDIR}/cross-seed:/torrents
```
And my config will be like that
```
	torrentDir: "/torrents",
	outputDir: "/cross-seeds",
```

**Comment by mmgoodnow** [comment]
> Hi, managed to fix this by adding
> 
> ```
> 
>     environment:
> 
>       - PGID=${PGID}
> 
>       - PUID=${PUID}
> 
> ```
> 
> Your configuration is valid!
> 
> Server is running on port 2468, ^C to stop.
> 
> 
> 
> I have no idea why not `user: 1000:1000`
> 
> 
> 
> I can also see the `cross-seed` tag added in `qbittorrent` but I cannot see anything added in the `qbittorrent`. Am I missing anything?
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/65309559/117067758-531f8500-ad22-11eb-88d3-853af8ccc578.png)
> 
> 
> 
> Thanks
> 
> 

That means that cross-seed is running as root, which is why it's no longer getting permission denied.

**Comment by mmgoodnow** [comment]
> Apologise when you say set the volume mapping you mean like that?
> 
> ```
> 
>     volumes:
> 
>       - /etc/localtime:/etc/localtime:ro
> 
>       - ${DOCKERCONFDIR}/cross-seed:/config
> 
>       - ${DOCKERCONFDIR}/cross-seed:/cross-seeds
> 
>       - ${DOCKERCONFDIR}/cross-seed:/torrents
> 
> ```
> 
> And my config will be like that
> 
> ```
> 
> 	torrentDir: "/torrents",
> 
> 	outputDir: "/cross-seeds",
> 
> ```

Basically, yes - but you'll want separate folders most likely.

**Comment by mmgoodnow** [comment]
You will also need to fill the torrentDir with all of your torrent files that you want to match against. It will use them for de-duplication as well so as not to download a torrent you already have.

**Comment by ajhorvath20** [comment]
So that's the changes I made 
```
  cross-seed:
    image: mmgoodnow/cross-seed
    hostname: ${DOCKERHOSTNAME}
    ports:
      - 2468:2468
    container_name: cross-seed
    environment:
      - PGID=${PGID}
      - PUID=${PUID}
      - TZ=${TZ}
    restart: unless-stopped
    command: daemon
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - ~/.config/appdata/cross-seed:/config
      - /mnt/data/.torrents:/cross-seeds
      - /mnt/qnap/.torrents:/torrents
```
And thats my `config.js`
```
module.exports = {
        configVersion: 1,

        jackettServerUrl: "http://192.168.1.167:9117",
        jackettApiKey: "apikey",
        trackers: ["oink", "tehconnection"],
        torrentDir: "/torrents",
        outputDir: "/cross-seeds",
        includeEpisodes: false,
        searchAll: false,
        action: "save",
        rtorrentRpcUrl: undefined,
        qbittorrentUrl: "http://username:password@192.168.1.167:8080",
};
```
That's my logs
```
~$ docker logs cross-seed 
Configuration file already exists.
Validating your configuration...
Your configuration is valid!
Server is running on port 2468, ^C to stop.
```
And the logs from `jackett`
```
today at 04:03:28  05-05 04:03:28 Info Manual search in mytracker for bscdjpssadadiadsasdddsnimcdqsanc => Found 0 releases (from cache) 
today at 04:06:00  05-05 04:06:00 Info Manual search in mytracker for bscdjpssadadiadsasdddsnimcdqsanc => Found 0 releases (from cache) 
today at 04:06:00  05-05 04:06:00 Info Manual search in mytracker for bscdjpssadadiadsasdddsnimcdqsanc => Found 0 releases (from cache) 
today at 04:08:22  05-05 04:08:22 Info Manual search in mytracker for bscdjpssadadiadsasdddsnimcdqsanc => Found 0 releases (from cache) 
today at 04:08:22  05-05 04:08:22 Info Manual search in mytracker for bscdjpssadadiadsasdddsnimcdqsanc => Found 0 releases (from cache) 
```
These are the `cross-seed` output from `qbittorrent`

![image](https://user-images.githubusercontent.com/65309559/117093441-d8259100-ad58-11eb-8faf-20bbc8f10d7f.png)

But still I am not seeing any cross-seedings in qbittorrent or .torrent files in the output directory. So I am lost now :)

**Comment by mmgoodnow** [comment]
It looks like you're trying to run Daemon mode. It hasn't been tested with qBittorrent yet (see #49). If you know what you're doing, I wouldn't mind some help testing. However, I think there might be some other confusion as well.

[Here](https://github.com/mmgoodnow/cross-seed/wiki/Daemon-Mode) are the docs for daemon mode. This might not do what you want it to do. The way daemon mode works is that it listens for YOU to send it requests to search for a given torrent. If you don't send it any HTTP requests, it won't do anything.

If that is what you want, then you'll need to set up your torrent client to send those requests on torrent completion, with `curl` and the `Run program on torrent completion` setting in your client. 

If that's not what you're looking for, you'll want to use search mode, which runs in a one-off fashion. 

Either way (daemon or search), you need to make sure that the `torrents` directory is filled with all of the torrents you're looking to cross-seed.

**Comment by ajhorvath20** [comment]
Thank you so much @mmgoodnow for detailed response and it really helped. I just changed the mode to `search` and it started working. 
```
Validating your configuration...
Your configuration is valid!
Found 9 torrents, 2 suitable to search for matches
```
If I understand correctly `cross-seed` should add the torrents into `qbittorrent` and `force-recheck`?

I will also try `daemon` mode because I think that's the way to move forward. Definitely I am happy to test, I am going to test it what you posted above and update you with the output.

Thanks again

**Comment by mmgoodnow** [comment]
you'll need "action": "inject" in your config for qBittorrent injection.

**Comment by ajhorvath20** [comment]
That works great. Awesome. Is it possible to set the permission right when I create an image. These are the permissions I have inside container currently 
```
drwxr-xr-x 1 root root  4.0K Apr 15 07:14 .
drwxr-xr-x 1 root root  4.0K Apr 14 06:36 ..
drwxr-xr-x 1 root staff 4.0K Apr 15 07:13 dist
drwxr-xr-x 1 root staff  12K Apr 15 07:13 node_modules
-rw-r--r-- 1 root staff 158K Apr 15 07:14 package-lock.json
-rw-r--r-- 1 root staff 1.4K Apr 15 07:11 package.json
drwxr-xr-x 1 root staff 4.0K Apr 15 07:13 src
-rw-r--r-- 1 root staff  220 Apr 15 07:11 tsconfig.json
```
But I have to `chown` it manually otherwise container was unable to create any of the files due to permission issue

![image](https://user-images.githubusercontent.com/65309559/117192233-604b7b00-add9-11eb-846c-c61a9e87c271.png)

And one more thing, is it possible for `cross-seed` to search and inject without the need of .torrent files?

Thanks

**Comment by mmgoodnow** [comment]
It looks like you're still running `cross-seed` as root. The PUID and PGID environment vars are a linuxserver.io convention that `cross-seed` doesn't observe. Are you saying that you're ***still*** getting permission errors running cross-seed as root?

AFAIK Docker always auto-creates directories as root. To avoid this you need to `chown` each directory that gets auto-created, but only once. After that you should be all set. Alternatively you can make the directories yourself beforehand. 

You mentioned machine A and machine B, are these just separate docker containers or are they fully separate computers? Hopefully it's the former. 

Cross-seed does need full read access to the torrent files. You should be able to find these in a `backup` subfolder of your qBittorrent config folder, I think, which you can then mount that folder to `/torrents` in the cross-seed container as read only with Docker.

**Comment by ajhorvath20** [comment]
This is completely make sense and I will change it to `user: 1000:1000`. Yes I manually created a folder and chown it and then added config.ts manually and create an image. All worked fine. I also mounted two folders 
```
      - ${DOCKERCONFDIR}/qbittorrent/data/BT_backup:/input:ro
      - ${DOCKERSTORAGEDIR}/data/.torrents:/output:rw
```
Only thing I am not understanding is some of the torrents (not all) are coming up as missing

![image](https://user-images.githubusercontent.com/65309559/117198225-a8ba6700-ade0-11eb-8825-a06b98c8e391.png)

Not sure why? Apologise for so many questions.

Thanks

**Comment by mmgoodnow** [comment]
> I also mounted two folders, input and output instead of torrents and cross-seed (if i am wrong please correct me to stay with torrents and cross-seeds)

That should be fine as long as you set `/input` and `/output` in your config file.

**Comment by mmgoodnow** [comment]
On the torrent with missing data, what changes do you need to make for it to start working correctly? Is it an issue with the category/save path?

**Comment by ajhorvath20** [comment]
Currently in Missing Files state location is set to `/storage/data/.torrents/` but to make it work I have to do `/storage/data/.torrents/movies/Movie.2021.1080p.BluRay`

**Comment by mmgoodnow** [comment]
man, some of these qBittorrent path mismatches are a total mystery to me. It seems like the `save_path` property returned by the API isn't totally reliable. I'll look into that for you but it'll take a few days. 

Is there any pattern to the torrents that are showing up as missing?

**Comment by ajhorvath20** [comment]
I tried to look at the pattern but just cannot find the exact reason but I attached the image and I hope that helps but please let me know if you need more information and thank you so much for looking into this for me.

![image](https://user-images.githubusercontent.com/65309559/117204621-6dbc3180-ade8-11eb-955a-a7d5f3c6a8c1.png)

**Comment by mmgoodnow** [comment]
Looks like the problem is with category-less torrents. I'll have a test-fix up in a few hours.

**Comment by ajhorvath20** [comment]
> Looks like the problem is with category-less torrents. I'll have a test-fix up in a few hours.

Thank you. I will be up in in couple of hours too so I can retest.

**Comment by ajhorvath20** [comment]
> It looks like you're trying to run Daemon mode. It hasn't been tested with qBittorrent yet (see #49). If you know what you're doing, I wouldn't mind some help testing. However, I think there might be some other confusion as well.
> 
> [Here](https://github.com/mmgoodnow/cross-seed/wiki/Daemon-Mode) are the docs for daemon mode. This might not do what you want it to do. The way daemon mode works is that it listens for YOU to send it requests to search for a given torrent. If you don't send it any HTTP requests, it won't do anything.
> 
> If that is what you want, then you'll need to set up your torrent client to send those requests on torrent completion, with `curl` and the `Run program on torrent completion` setting in your client.
> 
> If that's not what you're looking for, you'll want to use search mode, which runs in a one-off fashion.
> 
> Either way (daemon or search), you need to make sure that the `torrents` directory is filled with all of the torrents you're looking to cross-seed.

I am not sure what I need to put it in `"name":"<torrent name here>"`?

`curl -XPOST http://cross-seed:2468/api/webhook -H 'Content-Type: application/json' --data '{"name":"<torrent name here>",outputDir:"/storage/data/.torrents",trackers:["oink","tehconnection"]}'`

**Comment by mmgoodnow** [comment]
I think this image might help: https://user-images.githubusercontent.com/40671162/67574235-0fe55e80-f708-11e9-8441-6a4f30d19de4.png

**Comment by mmgoodnow** [comment]
I didn't get around to fixing yesterday but am hoping to have one for you ASAP.

**Comment by ajhorvath20** [comment]
> I think this image might help: ![image](https://user-images.githubusercontent.com/40671162/67574235-0fe55e80-f708-11e9-8441-6a4f30d19de4.png)

Hmm I tried with this `curl -XPOST http://cross-seed:2468/api/webhook -H 'Content-Type: application/json' --data '{"name":"%N"}'` but I am getting these outputs

qBittorrent
```
today at 20:03:07    % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
today at 20:03:07                                   Dload  Upload   Total   Spent    Left  Speed
today at 20:05:07  
today at 20:05:07  curl: (52) Empty reply from server
today at 20:05:07    % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
today at 20:05:07                                   Dload  Upload   Total   Spent    Left  Speed
today at 20:05:08  
today at 20:05:08  curl: (52) Empty reply from server
```
cross-seed
```
today at 17:21:19  Configuration file already exists.
today at 17:21:19  Validating your configuration...
today at 17:22:55  Your configuration is valid!
today at 17:24:33  Server is running on port 2468, ^C to stop.
today at 20:03:07  (node:1) UnhandledPromiseRejectionWarning: TypeError: Cannot destructure property 'name' of 'parseData(...)' as it is undefined.
today at 20:03:07      at Server.handleRequest (/usr/src/app/dist/src/server.js:66:13)
today at 20:03:07      at processTicksAndRejections (internal/process/task_queues.js:97:5)
today at 20:03:07  (node:1) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
today at 20:03:07  (node:1) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by ajhorvath20** [comment]
Running in `search` mode works fine but I see so many requests. Is it possible to use an environment variable to delay/limit the search i.e. 
```
    environment:
      - SEARCH=60m
```
Or using something like `command: search delay 30m`

Thanks

**Comment by mmgoodnow** [comment]
Okay. The latest release https://github.com/mmgoodnow/cross-seed/releases/tag/v2.3.4 will have a slightly more descriptive error message for your web hook problem. 

Sending JSON with `curl` through a shell can be really fiddly due to special characters (`{`, `}`, `"`, `'`) being intercepted by the shell. I don't totally know how they're performing the substitution on the qBittorrent end either, which may be complicating things. I recommend using the other format: `curl http://cross-seed:2468/api/webhook --data-urlencode "name=%N"`.

At the very least, the new version will tell you what the raw request body looks like, so we can further narrow down the problem.

**Comment by ajhorvath20** [comment]
Two more example if that helps you out guys

![image](https://user-images.githubusercontent.com/65309559/117864444-b6fafe00-b28c-11eb-9f76-2fc9c8d6dad8.png)

![image](https://user-images.githubusercontent.com/65309559/117864605-e7db3300-b28c-11eb-8ed8-c89d310af40f.png)

**Comment by bobokun** [comment]
I think the issue could be what was suggested in the previous post,
![image](https://user-images.githubusercontent.com/12660469/117864137-78e8e480-b262-11eb-9838-c0c90f4e5222.png)
These files that are seeding don't have a category linked to it so the location is not being added correctly. 

If the torrent management mode is set to automatic, You shouldn't need to pass the save_path and the category because the category is tied to the save_path. Simply passing the category will be enough when adding the torrent.

So what I'm guessing is that it's ignoring the save_path because the category is set to cross-seed and torrent management mode is set to automatic so it's overwriting the save_path with the default one (not the one that you passed).

Try setting the autoTMM option to false when adding the torrent file.
https://github.com/qbittorrent/qBittorrent/wiki/WebUI-API-(qBittorrent-4.1)#add-new-torrent

**Comment by ajhorvath20** [comment]
> Okay. The latest release https://github.com/mmgoodnow/cross-seed/releases/tag/v2.3.4 will have a slightly more descriptive error message for your web hook problem.
> 
> Sending JSON with `curl` through a shell can be really fiddly due to special characters (`{`, `}`, `"`, `'`) being intercepted by the shell. I don't totally know how they're performing the substitution on the qBittorrent end either, which may be complicating things. I recommend using the other format: `curl http://cross-seed:2468/api/webhook --data-urlencode "name=%N"`.
> 
> At the very least, the new version will tell you what the raw request body looks like, so we can further narrow down the problem.

This works but I have this log in `jackett`

![image](https://user-images.githubusercontent.com/65309559/118107203-61326d00-b3d6-11eb-9563-01a27f370648.png)

and this one from `cross-seed` 
```
today at 08:43:09  Found .2002.1080p.Amazon.WEB-DL.DD+2.0.x264-QOQ.mkv on trackername - exists
today at 08:43:09  Found 1 torrents for .2002.1080p.Amazon.WEB-DL.DD+2.0.x264-QOQ.mkv
```
but nothing sent to `qbittorrent`

Thanks

**Comment by mmgoodnow** [comment]
`- exists` means that you already have the match it found, in your client. I have gotten another similar bug report (#64), though, so it may be due to that. Can you check if the release it claimed existed actually does exist?

Thanks for bearing with me through all of these BitTorrent bugs; rTorrent was so much easier to get working because I'm familiar with it 😅

**Comment by mmgoodnow** [comment]
https://github.com/mmgoodnow/cross-seed/releases/tag/v2.3.5 should *hopefully* 🤞 fix the missing files bug. Auto torrent management will no longer be turned on for cross-seeds injected into BitTorrent. FYI @Drazzilb08

**Comment by ajhorvath20** [comment]
Hi, I am still seeing these

![image](https://user-images.githubusercontent.com/65309559/118405975-36fcdb80-b672-11eb-8299-2b96d90a01e4.png)

![image](https://user-images.githubusercontent.com/65309559/118405865-aaeab400-b671-11eb-93f6-34405869f073.png)

Thanks

**Comment by mmgoodnow** [comment]
I'm going to close this issue out because it has become a bit of a mess. I'll be separately opening an issue for the missing files bug - are there any other bugs still outstanding from this thread? I believe we got the daemon mode stuff working, right?

**Comment by mmgoodnow** [comment]
#74

**Comment by ajhorvath20** [comment]
Yes `daemon` is working great. Thank you. I thought of doing something like that, removing this from my compose `command: daemon` and then I can run it manually by `docker exec -it cross-seed cross-seed` 

```
root@test:/usr/src/app# ls -alh
total 196K
drwxr-xr-x 1 root root  4.0K May 13 23:37 .
drwxr-xr-x 1 root root  4.0K May 13 23:35 ..
drwxr-xr-x 1 root staff 4.0K May 13 23:36 dist
drwxr-xr-x 1 root staff  12K May 13 23:36 node_modules
-rw-r--r-- 1 root staff 158K May 13 23:37 package-lock.json
-rw-r--r-- 1 root staff 1.4K May 13 23:34 package.json
drwxr-xr-x 1 root staff 4.0K May 13 23:36 src
-rw-r--r-- 1 root staff  220 May 13 23:34 tsconfig.json
```
I will give a go but leave this closed as its sorted with `daemon`.

Thanks again :) and much appreciated

**Comment by Davo1624** [comment]
I am now getting this error as well when manually running a search command via `docker exec cross-seed cross-seed search 

Here is my compose:
```
  cross-seed:
    command: daemon
    container_name: cross-seed
    depends_on:
      - qbittorrent
      - prowlarr
    image: mmgoodnow/cross-seed:latest
    networks:
      - media
    restart: unless-stopped
    user: ${PUID}:${PGID}
    volumes:
      - ${CONF_PATH}/cross-seed:/config
      - ${DATA_PATH}/torrents:/data/torrents
      - ${CONF_PATH}/qbittorrent-wg/data/BT_backup:/input
```
When running in daemon mode the inject function works just fine, it is only when I try to invoke a search that I get the following error:
```
fs.js:1013
  handleErrorFromBinding(ctx);
  ^

Error: EACCES: permission denied, mkdir 'undefined'
    at Object.mkdirSync (fs.js:1013:3)
    at main (file:///usr/src/app/dist/pipeline.js:141:8)
    at async Command.<anonymous> (file:///usr/src/app/dist/cmd.js:136:9)
    at async Command.parseAsync (/usr/src/app/node_modules/commander/lib/command.js:923:5)
    at async file:///usr/src/app/dist/cmd.js:147:1 {
  errno: -13,
  syscall: 'mkdir',
  code: 'EACCES',
  path: 'undefined'
}
```
I have verified the `/config`, `/data/torrents` and `/input` directories all exist within the container

**Comment by mmgoodnow** [comment]
@Davo1624 what does your config file look like?

**Comment by Davo1624** [comment]
> @Davo1624 what does your config file look like?

Figured out the issue, for output dir in the config I had `undefined` and that was causing the error.
